### ISSUE TITLE
Possible overflow in typeobject.c:tail_contains

### ISSUE THREAD

**Issue by federicovalenso** [python/cpython#126862]
# Bug report

### Bug description:

[whence+1](https://github.com/python/cpython/blob/c0f045f7fd3bb7ccf9828f4bfad55347d097fd41/Objects/typeobject.c#L2867C14-L2867C22) could lead to overflow for large value of _whence_. I think [changing type from](https://github.com/python/cpython/blob/c0f045f7fd3bb7ccf9828f4bfad55347d097fd41/Objects/typeobject.c#L2986C5-L2986C44) _int_ to _Py_ssize_t_  could fix the problem (_remain_ is [input parameter](https://github.com/python/cpython/blob/c0f045f7fd3bb7ccf9828f4bfad55347d097fd41/Objects/typeobject.c#L3016)):
```c
static int
pmerge(PyObject *acc, PyObject **to_merge, Py_ssize_t to_merge_size)
{
...
    remain = PyMem_New(Py_ssize_t, to_merge_size);
```

### CPython versions tested on:

3.11

### Operating systems tested on:

Linux

### Linked PRs
* gh-127523
* gh-128699
* gh-128700

**Comment by picnixz** [comment]
This one is used in

```C
        candidate = PyTuple_GET_ITEM(cur_tuple, remain[i]);
        for (j = 0; j < to_merge_size; j++) {
            PyObject *j_lst = to_merge[j];
            if (tail_contains(j_lst, remain[j], candidate))
                goto skip; /* continue outer loop */
        }
```

and for the MRO resolution. Unless we have a VERY huge list of parent classes, I don't think we would hit the overflow.

**Comment by picnixz** [comment]
cc @JelleZijlstra

**Comment by federicovalenso** [comment]
@picnixz , should I try to overflow inheritance list? :) Or is there already protection against this?

**Comment by picnixz** [comment]
Yes, if you can make it happen! (I don't know whether there's a protection)

**Comment by picnixz** [comment]
I'll close this one as completed since we changed `int` to `Py_ssize_t` for cleanliness (well, `whence + 1` could still overflow, but this is in the infeasible realm)
